esp-hal/src/aes/mod.rs

```diff
@@ -595,6 +684,542 @@ pub mod dma {
+            if processed_bytes < item.buffers.input.len() {
+                let mut temp_item = AesOperation {
+                    mode: item.mode,
+                    cipher_mode: item.cipher_mode,
+                    buffers: unsafe { item.buffers.byte_add(processed_bytes) },
+                    key: item.key.copy(),
+                };
+
+                // If there is enough data for the DMA, set it up again:
+                let dma_processed = self.process_with_dma(&mut temp_item);
+                if dma_processed > 0 {
+                    self.state = DmaState::WaitingForDma {
+                        processed_bytes: processed_bytes + dma_processed,
+                    };
+                    return Some(Poll::Pending);
+                } else {
+                    // Otherwise, process the remaining data with CPU.
+                    self.process_with_cpu(&mut temp_item);
+                }
+            }
+
+            Some(Poll::Ready(Status::Completed))
+        }
+
+        fn process_with_cpu(&mut self, work_item: &mut AesOperation) {
+            let driver = unwrap!(self.driver.as_mut());
+            driver.aes.process_work_item(work_item);
+        }
+
+        fn process_with_dma(&mut self, work_item: &mut AesOperation) -> usize {
+            let (mut input_buffer, data_len) = unsafe {
+                prepare_for_tx(
+                    &mut self.input_descriptors,
+                    work_item.buffers.input,
+                    BLOCK_SIZE,
+                )
+            };
+
+            let number_of_blocks = data_len / BLOCK_SIZE;
+            if number_of_blocks == 0 {
+                // DMA can't do anything.
+                return 0;
+            }
+
+            let (mut output_buffer, rx_data_len) = unsafe {
+                prepare_for_rx(
+                    &mut self.output_descriptors,
+                    #[cfg(psram_dma)]
+                    &mut self.unaligned_data_buffers,
+                    // Truncate data based on how much the TX buffer can read.
+                    NonNull::slice_from_raw_parts(work_item.buffers.output.cast::<u8>(), data_len),
+                )
+            };
+
+            debug_assert_eq!(rx_data_len, data_len);
+
+            let driver = unwrap!(self.driver.as_mut());
+
+            let cipher_state = unsafe { work_item.cipher_mode.as_ref() };
+            cipher_state.write_state(driver);
+
+            let mode = cipher_state.hardware_operating_mode(work_item.mode, &work_item.key);
+            // This unwrap is fine, this function is not called for modes of operation not supported
+            // by the hardware.
+            let cipher_mode = unwrap!(cipher_state.hardware_cipher_mode());
+
+            unwrap!(driver.start_dma_transfer(
+                number_of_blocks,
+                &mut output_buffer,
+                &mut input_buffer,
+                mode,
+                cipher_mode,
+                &work_item.key
+            ));
```

Comment:
You can just call `process` here. You already have `Option<AesDma>`, making it an enum won't complicate things too much.
I think this is important for dogfooding the existing APIs and serves as an example of how users can wrap up the low-level APIs.

(I imagine this is probably your least favorite comment of mine 😂)